### PR TITLE
libroach: bump up rocks write-backpressure triggers significantly

### PR DESCRIPTION
**Pull request by dt**
We belive these triggers are unlikely to be hit in normal operation.
However they are easy to hit during bulk ingestion of SSTs, as that can quickly
add large amounts of data or numbers of files to any given level. When Rocks
assumes levels got that way via regular writes getting ahead of the compactor,
these triggers then start slowing those writes, affecting foreground traffic
that, in reality, is not what hit those triggers.

We considered dynamically raising these only durion bulk ingestion, but given
that we don't believe they are currently being hit, simply raising them should
achieve the same effect without any added complexity.

Release note (performance improvement): avoid rocksdb slowing down write traffic during bulk-ingestions.

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
I don't have a strong objection to this change. I'm curious to know what @ajkr thinks. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36424#-:-LbTKLVf3-LLyeIH6Q-4:bsiphou)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @petermattis)

**Comment by ajkr** [comment]
My understanding is we're assuming (1) the ingested files don't overlap the full keyspace; and (2) flushed L0 files, which do cover the full keyspace, cannot accumulate to trigger stalling. If either of these assumptions don't hold, then iteration can involve seeking/scanning in hundreds of files.

I don't particularly like (2) since I hope we will soon reach the point where normal writes can cause compaction debt to accumulate. But this feels fine until that day comes.

**Comment by dt** [comment]
I'm thinking of this specifically for 19.1 -- in 19.2 I want to further explore dynamically tuning for bulk-ingest vs regular work, so I think we'd be easily able to set these back to more aggressive defaults without worrying about them impacting bulk ops... but for 19.1, this felt like the simplest change.

**Comment by tbg** [comment]
In conjunction with backpressuring ingestion based on size of L0, is this still necessary? Large L0 is problematic in its own right.

**Comment by dt** [comment]
There's no back pressure on `IngestExternalFile`, and I wasn't planning on backporting anything to change that for 19.1. This was judged to be the simpler, less scary fix.

And while many files in L0 is indeed bad in its own right, that is particularly true when that is many files *than span the whole key span* i.e because they're from flushed memtables. That is assumed by the default compaction/back pressure settings. During ingest however, we might add 20 files to L0 that actually all only span some small, contained key span, and that span likely is seeing no or only a small fraction of the traffic to rocks, so they actually aren't going to impact overall traffic the the same way 20 full spanning L0 files would, right?

**Comment by petermattis** [comment]
> And while many files in L0 is indeed bad in its own right, that is particularly true when that is many files than span the whole key span i.e because they're from flushed memtables. That is assumed by the default compaction/back pressure settings. During ingest however, we might add 20 files to L0 that actually all only span some small, contained key span, and that span likely is seeing no or only a small fraction of the traffic to rocks, so they actually aren't going to impact overall traffic the the same way 20 full spanning L0 files would, right?

The current in memory organization of L0 tables is from oldest to newest, which means that we'll always consider all of the tables, even if it is to simply skip over them because the key we're reading lies outside their bounds. This organization makes sense if there are a handful of tables in L0 which mostly overlap. There is almost certainly a better organization structure if L0 can contain dozens or hundreds of tables which have minimal overlap.

Another thought is that the number of files in L0 is used to approximate the number of tables that have to be consulted. If we had a better in-memory structure for L0 files we could actually measure the max-overlap and use that as a write back pressure signal. Unfortunately, this all gets complicated with how iterators work, but certainly worth keeping in mind as a possibility for the future.

**Comment by dt** [comment]
Is there consensus on if we should/should not merge this?

**Comment by tbg** [comment]
Looks like it, though I'll defer to @petermattis || @ajkr. I also assume that you've run some experiments verifying this actually does "good things". And if we do accumulate lots of L0 (for example violating 2) in @ajkr's post) will we see it? I'm just not super confident that we weren't hitting write slowdowns before, it's not like we have good awareness for this sort of thing unless it's really pathological.

**Comment by petermattis** [comment]
I'm fine with merging this. @ajkr Even if the ingested files do overlap and impact reads significantly, when would we actually want to stop writes? Two additional observations to consider here:
* We fsync sstables before ingesting them so there is a limit on the ingestion bandwidth (though not on the number of tables if they are small).
* CRDB writes involve performing reads, so there is already natural backpressure on the speed of writes as the number of L0 tables grows and thus slows down reads.

**Comment by petermattis** [comment]
> I'm just not super confident that we weren't hitting write slowdowns before, it's not like we have good awareness for this sort of thing unless it's really pathological.

If we're worried about choosing the wrong setting here, we could make this a cluster setting and plumb down allowing this value to be changed dynamically (RocksDB supports that).

**Comment by dt** [comment]
If we're going to set it dynamically, I'd rather just raise and lower it _during bulk ingestion_. Indeed, I was thinking I'd do something like that on master later so this patch is mostly just with 19.1 backport in mind.

**Comment by ajkr** [comment]
>  @ajkr Even if the ingested files do overlap and impact reads significantly, when would we actually want to stop writes?

Well in the normal non-ingestion case it'll only do this after the gradual slowdown of writes fails to reduce compaction debt, which is probably rare. But I guess there has to be a limit somewhere otherwise space-amp and read-amp can increase indefinitely.

In the ingestion case it's more confusing because the gradual slowdown doesn't apply to ingested files, so it's easy to ingest a bunch and go straight from normal writing mode to stopped writes mode. I liked @dt's other changes to apply backpressure ourselves to prevent this case. After that it should be the case again that stopping writes only happens in rare cases where compaction has fallen farther and farther behind over a long period of time.

**Comment by petermattis** [comment]
> Well in the normal non-ingestion case it'll only do this after the gradual slowdown of writes fails to reduce compaction debt, which is probably rare. But I guess there has to be a limit somewhere otherwise space-amp and read-amp can increase indefinitely.
>
> In the ingestion case it's more confusing because the gradual slowdown doesn't apply to ingested files, so it's easy to ingest a bunch and go straight from normal writing mode to stopped writes mode. I liked @dt's other changes to apply backpressure ourselves to prevent this case. After that it should be the case again that stopping writes only happens in rare cases where compaction has fallen farther and farther behind over a long period of time.

Agreed that backpressuring at a higher level is desirable. The purpose of this PR is to raise the slowdown and stop limits presuming that they should never be hit. The evidence so far is that hitting the stop writes limit is dramatic and undesirable and we'd be better served with the gradual slowdown of more L0 files causing slower reads. The way I look at this is defense-in-depth.

**Comment by ajkr** [comment]
> > Well in the normal non-ingestion case it'll only do this after the gradual slowdown of writes fails to reduce compaction debt, which is probably rare. But I guess there has to be a limit somewhere otherwise space-amp and read-amp can increase indefinitely.
> > In the ingestion case it's more confusing because the gradual slowdown doesn't apply to ingested files, so it's easy to ingest a bunch and go straight from normal writing mode to stopped writes mode. I liked @dt's other changes to apply backpressure ourselves to prevent this case. After that it should be the case again that stopping writes only happens in rare cases where compaction has fallen farther and farther behind over a long period of time.
> 
> Agreed that backpressuring at a higher level is desirable. The purpose of this PR is to raise the slowdown and stop limits presuming that they should never be hit. The evidence so far is that hitting the stop writes limit is dramatic and undesirable and we'd be better served with the gradual slowdown of more L0 files causing slower reads. The way I look at this is defense-in-depth.

Makes sense. As long as the stop limit is still below the absolute worst read- and space-amp we can tolerate while still wanting to keep the DB alive, then LGTM.

Probably one day we should measure where is the crossover point (# of L0 files) where reads slow down enough that we'd prefer to throttle foreground writes and file ingestions. But I don't think it's relevant until we make throttling fairly apply to both foreground writes and file ingestions.

**Review by ajkr — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36424#-:-LbZwuRX91txTRwhZbsb:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr)

**Comment by dt** [comment]
TFTRs!

bors r+

I'll brush up the backpressure PR for ingestions (#34258) and we can decide if we want to also merge that -- I suspect we will.

I still think that in the longer term we still want to do something fancier to tune for ingest, ala #35236, since we're still seeing some really excessive compaction numbers in some backfills, so when we do look at that, maybe these limits will go back to more conservative thresholds? I donno.

**Pull request merged**
